Makefiles were modified to enable another shared folder for QEMU, to …

awerv opened the pull request:
…preserve the content of the secure storage between reboots.

Usage: set QEMU_PPS_ENABLE=y and adjust QEMU_PPS_HOST_DIR. It also
requires QEMU_VIRTFS_ENABLE to be set to "y".

Also added a buildroot post-script which appends lines to /etc/fstab,
so shared directories can be automatically mounted if
QEMU_VIRTFS_AUTOMOUNT and QEMU_PPS_AUTOMOUNT are set to "y".

Signed-off-by: Roland Nagy <rnagy@xmimx.tk>

etienne-lms commented:
This is nice stuff. @awerv can you update this P-R?

awerv commented:
Sorry, didn't have much time lately. Most of the requested fixes are done already. Hopefully next week I'll have time to finish this.

jforissier commented:
@awerv that's OK, next week can still make it in the next release (3.10.0), the week after will be code freeze however.

jforissier commented:
Removing the stale label and reopening.

awerv commented on common.mk:
```diff
@@ -32,9 +32,51 @@ CCACHE ?= $(shell which ccache) # Don't remove this comment (space is needed)
+QEMU_PPS_HOST_DIR		?= $(ROOT)
+
+# check to avoid virtfs = n and pps = y
+ifeq ($(QEMU_VIRTFS_ENABLE),n)
```
All the ifeq (*,n) were replaced to ifneq (*,y)

awerv commented on common.mk:
```diff
@@ -32,9 +32,51 @@ CCACHE ?= $(shell which ccache) # Don't remove this comment (space is needed)
+endif
+endif
+
+
```
Removed the double empty lines

awerv commented on common.mk:
```diff
@@ -203,6 +245,10 @@ BR2_PACKAGE_OPTEE_TEST_EXT_SITE ?= $(OPTEE_TEST_PATH)
 BR2_PACKAGE_STRACE ?= y
 BR2_TARGET_GENERIC_GETTY_PORT ?= $(if $(CFG_NW_CONSOLE_UART),ttyAMA$(CFG_NW_CONSOLE_UART),ttyAMA0)
 
+ifeq ($(BR2_ROOTFS_POST_BUILD_SCRIPT),$(ROOT)/build/br-ext/board/qemu/post-build.sh)
```
I wasn't sure if I can put the args into `qemu_v8.mk` since all the three used variables are defined in `common.mk`, bit it seems to work fine. Also added it to `qemu.mk`.

awerv commented on qemu_v8.mk:
```diff
@@ -14,6 +14,7 @@ override COMPILE_S_KERNEL  := 64
 TF_A_TRUSTED_BOARD_BOOT ?= n
 
 BR2_ROOTFS_OVERLAY = $(ROOT)/build/br-ext/board/qemu/overlay
+BR2_ROOTFS_POST_BUILD_SCRIPT = $(ROOT)/build/br-ext/board/qemu/post-build.sh
```
Added.

awerv commented on br-ext/board/qemu/post-build.sh:
```diff
@@ -0,0 +1,31 @@
+echo ""
+
+if [[ $VIRTFS_AUTOMOUNT == "y" ]];
+then
```
Folded.

awerv commented on br-ext/board/qemu/post-build.sh:
```diff
@@ -0,0 +1,31 @@
+fi
+
+if [[ $PSS_AUTOMOUNT == "y" ]];
+then
```
Folded.

etienne-lms replied:
Please address Jens' comments.

awerv commented:
I implemented the requested fixes, sorry for doing it so slowly. I hope you still have time to review it, and in case it's necessary, I'll try to fix further issues before next week's code freeze.

etienne-lms commented on br-ext/board/qemu/post-build.sh:
```diff
@@ -0,0 +1,31 @@
+echo "****************************************************************"
+echo ""
+echo "QEMU-specific post-build script invoked"
+echo ""
```
can remove this debug trace.

etienne-lms commented on br-ext/board/qemu/post-build.sh:
```diff
@@ -0,0 +1,31 @@
+#! /bin/bash
```
I think a copyright notice would be nice (despite other shell scripts don't have).

awerv replied:
Where can I find an example of such a copyright notice?

etienne-lms replied:
```bash
#! /bin/bash
# SPDX-License-Identifier: BSD-2-Clause
#
# Copyright (c) 2020, ...
```

etienne-lms commented on br-ext/board/qemu/post-build.sh:
```diff
@@ -0,0 +1,31 @@
+TARGETDIR="$1"
+VIRTFS_AUTOMOUNT="$2"
+VIRTFS_MOUNTPOINT="$3"
+PSS_AUTOMOUNT="$4"
```
should fail on missing or invalid data, no?

etienne-lms commented on common.mk:
```diff
@@ -32,9 +32,47 @@ CCACHE ?= $(shell which ccache) # Don't remove this comment (space is needed)
+# # Set QEMU_PPS_ENABLE to 'y' and adjust QEMU_PPS_HOST_DIR
+# # Then in QEMU, run:
+# # $ mount -t 9p -o trans=virtio secure /data/tee
+# # Or enable automount
```
suggest to replace `autmount` with `QEMU_VIRTFS_AUTOMOUNT`.

etienne-lms commented on common.mk:
```diff
@@ -32,9 +32,47 @@ CCACHE ?= $(shell which ccache) # Don't remove this comment (space is needed)
+QEMU_PPS_ENABLE			?= n
+QEMU_PPS_HOST_DIR		?= $(ROOT)
+
+# check to avoid virtfs = n and pps = y
```
minor: s/check/Check/
suggest to remove comment and use:
```makefile
ifeq ($(QEMU_PPS_ENABLE),y)
    $(call force,QEMU_VIRTFS_ENABLE,y,required by QEMU_PPS_ENABLE)
endif

etienne-lms commented on common.mk:
```diff
@@ -32,9 +32,47 @@ CCACHE ?= $(shell which ccache) # Don't remove this comment (space is needed)
+ifeq ($(QEMU_PSS_ENABLE),y)
+$(call force,QEMU_VIRTFS_ENABLE,y,required by QEMU_PSS_ENABLE)
+endif
+endif
```
I think you can simplify here.
`cal force...` already provides a useful description what's done so no need for comment.
`ifeq ($-QEMU_VIRTFS_ENABLE),y)` not needed.
```patch
-# Check to avoid virtfs = n and pps = y
-ifneq ($(QEMU_VIRTFS_ENABLE),y)
ifeq ($(QEMU_PSS_ENABLE),y)
$(call force,QEMU_VIRTFS_ENABLE,y,required by QEMU_PSS_ENABLE)
endif
-endif
```
And same below:
```patch
-ifneq ($(QEMU_VIRTFS_ENABLE),y)
ifeq ($(QEMU_VIRTFS_AUTOMOUNT),y)
-$(error Virtfs automount is enabled while virtfs disabled; enable virtfs or disable automount)
+$(call force,QEMU_VIRTFS_ENABLE,y,required by QEMU_VIRTFS_AUTOMOUNT)
endif
-endif

-ifneq ($(QEMU_PSS_ENABLE),y)
ifeq ($(QEMU_PSS_AUTOMOUNT),y)
-$(error Persistent secure storage automount is enabled while persistent secure storage disabled; enable persistent secure storage or disable automount)
+$(call force,QEMU_PSS_ENABLE,y,required by QEMU_PSS_AUTOMOUNT)
endif
-endif
```

awerv commented:
Fixed the commit description and added the `Reviewed-by` tag.

etienne-lms commented on common.mk:
```diff
@@ -32,9 +32,40 @@ CCACHE ?= $(shell which ccache) # Don't remove this comment (space is needed)
+QEMU_PSS_HOST_DIR		?= $(ROOT)
+
+ifeq ($(QEMU_PSS_ENABLE),y)
+$(call force,QEMU_VIRTFS_ENABLE,y,required by QEMU_PSS_ENABLE)
```
I'm sorry, I mixed up the build environments.
Actually macro `force` is something defined only in optee_os make script, not here in the **build.git** makefiles.

Yet, since this syntax looks nice, I think we could define the macro in this **common.mk** file:
```makefile
# Set a variable or error out if it was previously set to a different value
# The reason message (3rd parameter) is optional
# Example:
# $(call force,CFG_FOO,foo,required by CFG_BAR)
define force
$(eval $(call _force,$(1),$(2),$(3)))
endef

define _force
ifdef $(1)
ifneq ($($(1)),$(2))
ifneq (,$(3))
_reason := $$(_empty) [$(3)]
endif
$$(error $(1) is set to '$($(1))' (from $(origin $(1))) but its value must be '$(2)'$$(_reason))
endif
endif
$(1) := $(2)
endef
```

jforissier replied:
> I'm sorry, I mixed up the build environments.

Me too :-/

> Yet, since this syntax looks nice, I think we could define the macro in this common.mk file:

Agreed.

awerv replied:
Added it to the `defines, macros...` section of `common.mk`.

jforissier commented:
Actually there are ordering issues in `common.mk`. I am doing some tests and will post a patch shortly.

awerv commented:
> Actually there are ordering issues in `common.mk`. I am doing some tests and will post a patch shortly.

`force` doesn't work properly if used before it is defined, so it should be defined before `ifeq ($(QEMU_PSS_ENABLE),y)`. Any suggestions about where shall I define it?

etienne-lms commented:
Line 30 of **common.mk** looks a good place to me: almost at makefile top.

awerv commented:
> Line 30 of **common.mk** looks a good place to me: almost at makefile top.

Moved it up right after `CCACHE ?= ...`

jforissier commented:
Please see https://github.com/jforissier/optee_repo_build_files/commit/b41a5d9a614b0aa315c93b199b336a83232a3a45.

If you cherry-pick the above commit and squash with yours, then you may add:
```
Reviewed-by: Jerome Forissier <jerome@forissier.org>
Tested-by: Jerome Forissier <jerome@forissier.org>
```

awerv commented:
> Please see [jforissier@b41a5d9](https://github.com/jforissier/optee_repo_build_files/commit/b41a5d9a614b0aa315c93b199b336a83232a3a45).
> 
> If you cherry-pick the above commit and squash with yours, then you may add:
> 
> ```
> Reviewed-by: Jerome Forissier <jerome@forissier.org>
> Tested-by: Jerome Forissier <jerome@forissier.org>
> ```

Thank you, squashed them and added the `Tested-by` tag.
